pkg/apis/resource/validation/validation.go

```diff
@@ -17,172 +17,264 @@ limitations under the License.
 package validation
 
 import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+
 	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	apiresource "k8s.io/apimachinery/pkg/api/resource"
 	apimachineryvalidation "k8s.io/apimachinery/pkg/api/validation"
-	pathvalidation "k8s.io/apimachinery/pkg/api/validation/path"
 	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/apimachinery/pkg/util/sets"
-	utilvalidation "k8s.io/apimachinery/pkg/util/validation"
+	"k8s.io/apimachinery/pkg/util/validation"
 	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/apiserver/pkg/cel"
+	"k8s.io/apiserver/pkg/cel/environment"
+	dracel "k8s.io/dynamic-resource-allocation/cel"
 	corevalidation "k8s.io/kubernetes/pkg/apis/core/validation"
 	"k8s.io/kubernetes/pkg/apis/resource"
-	namedresourcesvalidation "k8s.io/kubernetes/pkg/apis/resource/structured/namedresources/validation"
 )
 
-// validateResourceDriverName reuses the validation of a CSI driver because
-// the allowed values are exactly the same.
-var validateResourceDriverName = corevalidation.ValidateCSIDriverName
+var (
+	// validateResourceDriverName reuses the validation of a CSI driver because
+	// the allowed values are exactly the same.
+	validateDriverName      = corevalidation.ValidateCSIDriverName
+	validateDeviceName      = corevalidation.ValidateDNS1123Subdomain
+	validateDeviceClassName = corevalidation.ValidateDNS1123Subdomain
```

Comment:
No strong opinion.

@klueska ?

Reply:
I have historically been using `gpu.nvidia.com` as the "default" class name (to mimic the default extended resource name of `nvidia.com/gpu`). Can I flip the question around -- is there a strong reason *not* to allow a subdomain?

Reply:
Allowing domains might indeed be useful, not just to mimic traditional behavior but also to avoid name clashes between classes defined by different vendors. We don't have any guidelines for classes, but if we had, then "name ends with a domain owned by the vendor" would make sense for classes that come straight from a vendor.

Reply:
This is more of a less-is-more question.  There's a whole lot of cargo-culting in validation, and using domains rather than labels has occasionally caused problems.  As long as we are doing it for a reason, I am OK with it.  For non-namespaces objects it seems less likely to be a problem than namespaced things.